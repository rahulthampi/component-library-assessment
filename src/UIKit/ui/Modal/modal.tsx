import * as React from 'react';
import * as DialogPrimitive from '@radix-ui/react-dialog';
import { X } from 'lucide-react';
import { VariantProps, cva } from 'class-variance-authority';

import { cn } from '../../../lib/utils';

const Root = DialogPrimitive.Root;

const Trigger = DialogPrimitive.Trigger;

const Portal = DialogPrimitive.Portal;

const Close = DialogPrimitive.Close;

const Overlay = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Overlay>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Overlay>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Overlay
    ref={ref}
    className={cn(
      'fixed inset-0 z-50 bg-background/80 backdrop-blur-sm data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0',
      className,
    )}
    {...props}
  />
));
Overlay.displayName = DialogPrimitive.Overlay.displayName;

const Header = ({ className, ...props }: React.HTMLAttributes<HTMLDivElement>) => (
  <div className={cn('flex flex-col space-y-1.5 text-center sm:text-left', className)} {...props} />
);
Header.displayName = 'DialogHeader';

const Footer = ({ className, ...props }: React.HTMLAttributes<HTMLDivElement>) => (
  <div
    className={cn('flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2', className)}
    {...props}
  />
);
Footer.displayName = 'DialogFooter';

const Title = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Title>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Title>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Title
    ref={ref}
    className={cn('text-lg font-semibold leading-none tracking-tight', className)}
    {...props}
  />
));
Title.displayName = DialogPrimitive.Title.displayName;

const Description = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Description>,
  React.ComponentPropsWithoutRef<typeof DialogPrimitive.Description>
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Description
    ref={ref}
    className={cn('text-sm text-muted-foreground', className)}
    {...props}
  />
));
Description.displayName = DialogPrimitive.Description.displayName;

const dialogContentVariants = cva(
  'fixed left-[50%] top-[50%] z-50 grid w-full translate-x-[-50%] translate-y-[-50%] gap-4 border bg-background p-6 shadow-md duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg md:w-full',
  {
    variants: {
      size: {
        sm: 'max-w-[400px]',
        md: 'max-w-[600px]',
        lg: 'max-w-[960px]',
        xl: 'max-w-[1248px]',
      },
    },
    defaultVariants: {
      size: 'md',
    },
  },
);

interface ContentProps
  extends React.ComponentPropsWithoutRef<typeof DialogPrimitive.Content>,
    VariantProps<typeof dialogContentVariants> {
  showClose?: boolean;
  closeIcon?: React.ReactNode;
}

const Content = React.forwardRef<React.ElementRef<typeof DialogPrimitive.Content>, ContentProps>(
  (
    {
      children,
      className,
      size,
      showClose = true,
      closeIcon = <X className="h-4 w-4" />,
      ...props
    },
    ref,
  ) => (
    <Portal>
      <Overlay />

      <DialogPrimitive.Content
        ref={ref}
        className={cn(dialogContentVariants({ size, className }), className)}
        data-testid="modal-content"
        {...props}
      >
        {children}

        {showClose ? (
          <Close className="absolute right-4 top-4 rounded-sm opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground">
            {closeIcon}
            <span className="sr-only">Close</span>
          </Close>
        ) : null}
      </DialogPrimitive.Content>
    </Portal>
  ),
);
Content.displayName = DialogPrimitive.Content.displayName;

export { type ContentProps };

export { Root, Trigger, Content, Header, Footer, Title, Description, Close };
