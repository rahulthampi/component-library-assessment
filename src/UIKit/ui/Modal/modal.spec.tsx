import { render, screen, fireEvent } from '@testing-library/react';
import { toMatchDiffSnapshot } from 'snapshot-diff';
import '@testing-library/jest-dom';

import * as Modal from './modal';
import { Button } from '../button';

expect.extend({ toMatchDiffSnapshot });

const Trigger = () => {
  return (
    <Modal.Trigger asChild>
      <Button variant="outline">Open Modal</Button>
    </Modal.Trigger>
  );
};

const Title = () => {
  return (
    <Modal.Header>
      <Modal.Title>Modal Title</Modal.Title>
    </Modal.Header>
  );
};

const Content = () => {
  return (
    <>
      <p className="text-t-sm">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse eget purus sagittis,
        egestas purus at, interdum nisi. Vestibulum varius dolor sem, non faucibus justo accumsan
        vitae. Nullam ut quam pretium nisi laoreet viverra vitae nec massa. Quisque volutpat
        vulputate dui, ac malesuada lectus. Suspendisse quis eros a tortor viverra commodo. Etiam
        non orci metus. Praesent et ex lorem. Curabitur ornare venenatis ornare. Orci varius natoque
        penatibus et magnis dis parturient montes, nascetur ridiculus mus. Suspendisse potenti.
      </p>

      <p className="text-t-sm">
        Vestibulum velit nisl, vehicula sed convallis vitae, malesuada id felis. Donec fringilla
        pretium justo et rutrum. Donec id ipsum orci. Pellentesque iaculis vehicula tellus quis
        rhoncus. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos
        himenaeos. Sed semper, purus eu porttitor interdum, mauris lacus pharetra arcu, et congue
        est neque a tortor. Nam sit amet neque quam. Phasellus est felis, placerat in orci et,
        lacinia convallis lorem. Duis mi lectus, placerat a velit a, auctor vehicula lectus.
      </p>
    </>
  );
};

const Footer = () => {
  return (
    <Modal.Footer>
      <Modal.Close asChild>
        <Button type="button" variant="secondary">
          Close Modal
        </Button>
      </Modal.Close>
    </Modal.Footer>
  );
};

describe('UIKit/Modal', () => {
  it('should render only the trigger and no modal content', () => {
    const rendered = render(
      <Modal.Root>
        <Trigger />

        <Modal.Content>
          <Title />

          <Content />

          <Footer />
        </Modal.Content>
      </Modal.Root>,
    );

    expect(rendered).toMatchSnapshot();

    expect(screen.getByRole('button')).not.toBeDisabled();

    expect(screen.queryByText('Modal Title')).not.toBeInTheDocument();
  });

  it.each([
    ['sm', 'max-w-[400px]'] as const,
    ['md', 'max-w-[600px]'] as const,
    ['lg', 'max-w-[960px]'] as const,
    ['xl', 'max-w-[1248px]'] as const,
  ])(
    'should open modal(of size: %s) and close the modal when Trigger and Close are clicked',
    (size, expectedClass) => {
      const { asFragment } = render(
        <Modal.Root>
          <Trigger />

          <Modal.Content size={size}>
            <Title />

            <Content />

            <Footer />
          </Modal.Content>
        </Modal.Root>,
      );

      const noContentRender = asFragment();

      const openButton = screen.getByText('Open Modal');

      // Click the open button to open the modal.
      fireEvent.click(openButton);

      expect(noContentRender).toMatchDiffSnapshot(asFragment());

      const closeButton = screen.getByText('Close Modal');

      expect(screen.getByText('Modal Title')).toBeInTheDocument();
      expect(closeButton).toBeInTheDocument();
      expect(screen.getByTestId('modal-content')).toHaveClass(expectedClass);

      // Click the close button to close the modal.
      fireEvent.click(closeButton);
      expect(screen.queryByText('Modal Title')).not.toBeInTheDocument();
    },
  );
});
