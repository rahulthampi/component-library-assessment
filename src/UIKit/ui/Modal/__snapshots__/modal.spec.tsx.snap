// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIKit/Modal should open modal(of size: lg) and close the modal when Trigger and Close are clicked 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <button
      aria-controls="radix-:r9:"
-     aria-expanded="false"
+     aria-expanded="true"
      aria-haspopup="dialog"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
-     data-state="closed"
+     data-state="open"
      type="button"
    >
      Open Modal
    </button>
  </DocumentFragment>"
`;

exports[`UIKit/Modal should open modal(of size: md) and close the modal when Trigger and Close are clicked 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <button
      aria-controls="radix-:r6:"
-     aria-expanded="false"
+     aria-expanded="true"
      aria-haspopup="dialog"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
-     data-state="closed"
+     data-state="open"
      type="button"
    >
      Open Modal
    </button>
  </DocumentFragment>"
`;

exports[`UIKit/Modal should open modal(of size: sm) and close the modal when Trigger and Close are clicked 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <button
      aria-controls="radix-:r3:"
-     aria-expanded="false"
+     aria-expanded="true"
      aria-haspopup="dialog"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
-     data-state="closed"
+     data-state="open"
      type="button"
    >
      Open Modal
    </button>
  </DocumentFragment>"
`;

exports[`UIKit/Modal should open modal(of size: xl) and close the modal when Trigger and Close are clicked 1`] = `
"Snapshot Diff:
- First value
+ Second value

  <DocumentFragment>
    <button
      aria-controls="radix-:rc:"
-     aria-expanded="false"
+     aria-expanded="true"
      aria-haspopup="dialog"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
-     data-state="closed"
+     data-state="open"
      type="button"
    >
      Open Modal
    </button>
  </DocumentFragment>"
`;

exports[`UIKit/Modal should render only the trigger and no modal content 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <button
        aria-controls="radix-:r0:"
        aria-expanded="false"
        aria-haspopup="dialog"
        class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
        data-state="closed"
        type="button"
      >
        Open Modal
      </button>
    </div>
  </body>,
  "container": <div>
    <button
      aria-controls="radix-:r0:"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2"
      data-state="closed"
      type="button"
    >
      Open Modal
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
